src/storage/store/block_engine.rs

```diff
@@ -0,0 +1,549 @@
+                    transaction.user_messages = vec![];
+                }
+
+                if transaction.system_messages.is_empty() && transaction.user_messages.is_empty() {
+                    return None;
+                } else {
+                    return Some(transaction);
+                }
+            })
+            .collect_vec();
+
+        for snapchain_txn in &mut snapchain_txns {
+            let account_root =
+                self.replay_snapchain_txn(&merkle_trie::Context::new(), &snapchain_txn, txn_batch)?;
+            snapchain_txn.account_root = account_root;
+        }
+
+        let (events, events_hash) = self.generate_block_events(height, timestamp, txn_batch);
+
+        self.metrics
+            .publish_transaction_counts(&snapchain_txns, ProposalSource::Propose);
+
+        let new_root_hash = self.trie.root_hash()?;
+
+        let result = ShardStateChange {
+            timestamp: timestamp.clone(),
+            new_state_root: new_root_hash.clone(),
+            transactions: snapchain_txns,
+            events_hash,
+            events,
+        };
+
+        Ok(result)
+    }
+
+    pub fn propose_state_change(
+        &mut self,
+        messages: Vec<MempoolMessage>,
+        height: Height,
+    ) -> ShardStateChange {
+        let now = std::time::Instant::now();
+        let mut txn = RocksDbTransactionBatch::new();
+
+        let timestamp = FarcasterTime::current();
+        let version = EngineVersion::version_for(&timestamp, self.network);
+        let state_change = if version.is_enabled(ProtocolFeature::WriteDataToShardZero) {
+            let result = self
+                .prepare_proposal(&mut txn, messages, height, &timestamp)
+                .unwrap();
+
+            self.trie.reload(&self.db).unwrap();
+            result
+        } else {
+            ShardStateChange {
+                events: vec![],
+                new_state_root: vec![],
+                timestamp: FarcasterTime::current(),
+                events_hash: vec![],
+                transactions: vec![],
+            }
+        };
+
+        let proposal_duration = now.elapsed();
+        self.metrics
+            .time_with_shard("propose_time", proposal_duration.as_millis() as u64);
+
+        self.metrics.count("propose.invoked", 1, vec![]);
+        state_change
+    }
+
+    fn replay_proposal(
+        &mut self,
+        txn_batch: &mut RocksDbTransactionBatch,
+        transactions: &[Transaction],
+        shard_root: &[u8],
+        events_hash: &Vec<u8>,
+        source: ProposalSource,
+        height: Height,
+        timestamp: &FarcasterTime,
+    ) -> Result<(), BlockEngineError> {
+        let now = std::time::Instant::now();
+        // Validate that the trie is in a good place to start with
+        match self.get_last_block() {
+            None => { // There are places where it's hard to provide a parent hash-- e.g. tests so make this an option and skip validation if not present
```

Comment:
I'm realizing we should probably only do this if we're in a test env (or devnet). Something for later.

Reply:
Added a comment